### PR TITLE
CB-17162: Show warning message when we are unable to read details of the Azure Marketplace image.

### PR DESCRIPTION
**Pull request by tiborpopovics**
At the beginning of the cluster provision, we're validating that the Terms and Conditions are accepted for the Azure Marketplace image.
To perform this validation we need the Microsoft.MarketplaceOrdering/offertypes/publishers/offers/plans/agreements/read permission.
Sometimes the customers are not providing this permission. After this change, we're sending an error message to the customer to make sure that the Terms and Conditions are accepted.

### CODE REVIEW

File: cloud-reactor/src/test/java/com/sequenceiq/cloudbreak/cloud/handler/ProvisionValidationHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+        verifyCloudConnectorCalls(request, cloudContext);
+        ArgumentCaptor<Event> argumentCaptor = ArgumentCaptor.forClass(Event.class);
+        verify(eventBus).notify(eq("VALIDATIONRESULT"), argumentCaptor.capture());
+        Event<ValidationResult> resultEvent = argumentCaptor.getValue();
+        assertTrue(resultEvent.getData().getWarningMessages().isEmpty());
+    }
+
+    @Test
+    void testAcceptShouldSendEventWhenAllValidationWereSuccessfulWhenTheValidatorThrowsWarningMessage() {
```
Comment (gergopapi2) [inline cloud-reactor/src/test/java/com/sequenceiq/cloudbreak/cloud/handler/ProvisionValidationHandlerTest.java:81]: I would make this name shorter and leave part  "WhenAllValidationWereSuccessful"

File: cloud-reactor/src/test/java/com/sequenceiq/cloudbreak/cloud/handler/ProvisionValidationHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        ValidationRequest request = createValidationRequest();
+        CloudContext cloudContext = request.getCloudContext();
+        mockCloudConnector(request, cloudContext);
+        doThrow(new CloudPlatformValidationWarningException(WARNING_MESSAGE, new Exception()))
+                .when(validator).validate(ac, request.getCloudStack());
+        underTest.accept(new Event<>(request));
```
Comment (gergopapi2) [inline cloud-reactor/src/test/java/com/sequenceiq/cloudbreak/cloud/handler/ProvisionValidationHandlerTest.java:87]: I would add a blank line above.

File: cloud-reactor/src/test/java/com/sequenceiq/cloudbreak/cloud/handler/ProvisionValidationHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Test
+    void testAcceptShouldSendFailureEventWhenTheValidationFailed() {
+        ValidationRequest request = createValidationRequest();
+        CloudContext cloudContext = request.getCloudContext();
+        mockCloudConnector(request, cloudContext);
+        CloudConnectorException exception = new CloudConnectorException("Error");
+        doThrow(exception).when(validator).validate(ac, request.getCloudStack());
+        underTest.accept(new Event<>(request));
```
Comment (gergopapi2) [inline cloud-reactor/src/test/java/com/sequenceiq/cloudbreak/cloud/handler/ProvisionValidationHandlerTest.java:105]: Please add a blank line above.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/marketplace/AzureImageTermsSignerService.java [hunk 4]
```diff
@@ -43,6 +48,10 @@ public boolean isSigned(String subscriptionId, AzureMarketplaceImage azureMarket
             AzureImageTerms azureImageTerms = azureRestOperationsService.httpGet(signUri, AzureImageTerms.class, token);
             LOGGER.debug("Image terms and conditions received for image {} is : {}", azureMarketplaceImage, azureImageTerms);
             return azureImageTerms.getProperties().isAccepted();
+        } catch (AzureRestResponseException e) {
+            String message = String.format(MISSING_READ_PERMISSION_MESSAGE_TEMPLATE, azureMarketplaceImage);
```
[the remark this answers is not in the document]
Comment (gergopapi2) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/marketplace/AzureImageTermsSignerService.java:52]: Also, is there anything specific in the AzureRestResponseException that helps find out that there is not enough permission to read the terms and conditions of the image?

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/ProvisionValidationHandler.java [hunk 5]
```diff
@@ -45,10 +49,16 @@ public void accept(Event<ValidationRequest> event) {
...
             CloudStack cloudStack = request.getCloudStack();
+            Set<String> warningMessages = new HashSet<>();
             for (Validator v : connector.validators(ValidatorType.ALL)) {
-                v.validate(ac, cloudStack);
+                try {
+                    v.validate(ac, cloudStack);
+                } catch (CloudPlatformValidationWarningException e) {
+                    LOGGER.warn("{} sent the following warning: {}", v.getClass().getSimpleName(), e.getMessage());
+                    warningMessages.add(e.getMessage());
+                }
```
Comment (gergopapi2) [inline cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/ProvisionValidationHandler.java:59]: The try-catch could go into a separate method.

File: common/src/main/resources/messages/messages.properties [hunk 6]
```diff
@@ -214,6 +214,7 @@ cluster.ccm.upgrade.finished=Cluster Connectivity Manager upgrade finished.
 cluster.kerberosconfig.validation.failed=Kerberos config validation failed with: {0}
 
 cluster.cloudconfig.validation.failed=Cloud config validation failed with the following errors: {0}
+cluster.provider.validation.warning=Some warnings were found during cloud platform validation: {0}
```
Comment (gergopapi2) [inline common/src/main/resources/messages/messages.properties:217]: I would reformulate to match the above error message:
"Cloud platform validation succeeded with the following warnings: {0}"

**Review by gergopapi2 — CHANGES_REQUESTED** [review]

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/marketplace/AzureImageTermsSignerService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tiborpopovics) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/marketplace/AzureImageTermsSignerService.java:52]: There is no specific in the exception. Only the message contains the `403 Forbidden` snippet and the details of the error.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/marketplace/AzureImageTermsSignerService.java [hunk 7]
```diff
@@ -30,6 +32,9 @@ public class AzureImageTermsSignerService {
     private static final String SIGN_URL_AZ_TEMPLATE = "https://management.azure.com/subscriptions/%s/providers/Microsoft.MarketplaceOrdering/offerTypes/" +
             "virtualmachine/publishers/%s/offers/%s/plans/%s/agreements/current?api-version=2015-06-01";
 
+    private static final String READ_PROBLEM_MESSAGE_TEMPLATE = "Failed to get the status of the Terms and Conditions for image %s. " +
```
Comment (pdarvasi) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/image/marketplace/AzureImageTermsSignerService.java:35]: Ideally, this should be in `messages.properties` instead

**Review by keyki — APPROVED** [review]

**Review by sodre90 — APPROVED** [review]

**Pull request merged**